nvse/nvse/NVSECompiler.cpp

```diff
@@ -0,0 +1,760 @@
+
+	return true;
+}
+
+void NVSECompiler::visitNVSEScript(const NVSEScript* nvScript) {
+	// Compile the script name
+	scriptName = nvScript->name.lexeme;
+
+	// Dont allow naming script the same as another form, unless that form is the script itself
+	auto comp = strcmp(scriptName.c_str(), originalScriptName);
+	if (resolveObjectReference(scriptName, false) && comp) {
+		throw std::runtime_error(std::format("Error: Form name '{}' is already in use.\n", scriptName.c_str()));
+	}
+
+	add_u32(0x1D);
+
+	for (auto& global_var : nvScript->globalVars) {
+		global_var->accept(this);
+	}
+
+	for (auto& block : nvScript->blocks) {
+		block->accept(this);
+	}
+}
+
+void NVSECompiler::visitBeginStatement(const BeginStmt* stmt) {
+	// Get opcode
+	auto beginInfo = mpBeginInfo[stmt->name.lexeme];
+
+	// OP_BEGIN
+	add_u16(0x10);
+
+	auto beginPatch = add_u16(0x0);
+	auto beginStart = data.size();
+
+	add_u16(beginInfo.opcode);
+
+	auto blockPatch = add_u32(0x0);
+
+	// Add param shit here?
+	if (stmt->param.has_value()) {
+		auto param = stmt->param.value();
+
+		add_u16(0xFFFF);
+		auto varStart = data.size();
+		auto varPatch = add_u16(0x0);
+
+		if (beginInfo.argType == ArgType::Int) {
+			add_u8('L');
+			add_u32(static_cast<uint32_t>(std::get<double>(param.value)));
+		}
+		else if (beginInfo.argType == ArgType::Ref) {
+			// Try to resolve the global ref
+			auto ref = resolveObjectReference(param.lexeme);
+			if (ref) {
+				auto refInfo = script->refList.GetNthItem(ref - 1);
+				printf("Resolved ref %s: %08x\n", param.lexeme.c_str(), refInfo->form->refID);
+				add_u8('R');
+				add_u16(ref);
+			} else {
+				throw std::runtime_error(std::format("Unable to resolve form '{}'.", param.lexeme));
+			}
+		}
+
+		set_u16(varPatch, data.size() - varStart);
+	}
+
+	set_u16(beginPatch, data.size() - beginStart);
+	auto blockStart = data.size();
+	stmt->block->accept(this);
+
+	// OP_END
+	add_u32(0x11);
+
+	set_u32(blockPatch, data.size() - blockStart);
+}
+
+void NVSECompiler::visitFnDeclStmt(FnDeclStmt* stmt) {
+	// OP_BEGIN
+	add_u16(0x10);
+
+	// OP_MODE_LEN
+	auto opModeLenPatch = add_u16(0x0);
+	auto opModeStart = data.size();
+
+	// OP_MODE
+	add_u16(0x0D);
+
+	// SCRIPT_LEN
+	auto scriptLenPatch = add_u32(0x0);
+
+	// BYTECODE VER
+	add_u8(0x1);
+
+	// arg count
+	add_u8(stmt->args.size());
+
+	// add args
+	for (auto i = 0; i < stmt->args.size(); i++) {
+		auto& arg = stmt->args[i];
+		auto type = getArgType(arg->type);
+		auto localIdx = addLocal(arg->name.lexeme, type);
+
+		add_u16(localIdx);
+		add_u8(type);
+	}
+
+	// NUM_ARRAY_ARGS, always 0
+	add_u8(0x0);
+
+	auto scriptLenStart = data.size();
+
+	// Patch mode len
+	set_u16(opModeLenPatch, data.size() - opModeStart);
+
+	// Compile script
+	compileBlock(stmt->body, false);
+
+	// OP_END
+	add_u32(0x11);
+
+	set_u32(scriptLenPatch, data.size() - scriptLenStart);
+}
+
+void NVSECompiler::visitVarDeclStmt(const VarDeclStmt* stmt) {
```

Comment:
Better to use const references than pointers here, i.e. `const VarDeclStmt& stmt` and then pass with `accept(*this)`